drivers/eeprom/eeprom_mchp_xec.c

```diff
@@ -6,25 +6,223 @@
+#define XEC_EEPROM_TRANSFER_SIZE_WRITE	XEC_EEPROM_PAGE_SIZE
+
+#define XEC_EEPROM_DELAY_US				500U
+#define XEC_EEPROM_XFER_COMPL_RETRY_COUNT		10U
+
+struct eeprom_xec_regs {
+	uint32_t mode;
+	uint32_t execute;
+	uint32_t status;
+	uint32_t intr_enable;
+	uint32_t password;
+	uint32_t unlock;
+	uint32_t lock;
+	uint32_t _reserved;
+	uint8_t buffer[XEC_EEPROM_PAGE_SIZE];
+};
 
 struct eeprom_xec_config {
-	uint32_t addr;
+	struct eeprom_xec_regs * const regs;
 	size_t size;
+#ifdef CONFIG_PINCTRL
+	const struct pinctrl_dev_config *pcfg;
+#endif
 };
 
+K_MUTEX_DEFINE(lock);
+
+static void eeprom_xec_execute_reg_set(struct eeprom_xec_regs * const regs,
+						uint32_t transfer_size, uint32_t command,
+						uint16_t eeprom_addr)
+{
+	uint32_t temp = command + (eeprom_addr & XEC_EEPROM_EXC_ADDR_MASK);
+
+	if (transfer_size != XEC_EEPROM_PAGE_SIZE) {
+		temp += (transfer_size << XEC_EEPROM_EXC_TRANSFER_SZ_BITPOS);
+	}
+	regs->execute = temp;
+}
+
+static void eeprom_xec_data_buffer_read(struct eeprom_xec_regs * const regs,
+					uint8_t transfer_size, uint8_t *destination_ptr)
+{
+	uint8_t count;
+
+	if (transfer_size <= XEC_EEPROM_PAGE_SIZE) {
+		for (count = 0; count < transfer_size; count++) {
+			*destination_ptr = regs->buffer[count];
+			destination_ptr++;
+		}
+	}
+}
+
+static void eeprom_xec_data_buffer_write(struct eeprom_xec_regs * const regs,
+							uint8_t transfer_size, uint8_t *source_ptr)
+{
+	uint8_t count;
+
+	if (transfer_size <= XEC_EEPROM_PAGE_SIZE) {
+		for (count = 0; count < transfer_size; count++) {
+			regs->buffer[count] = *source_ptr;
+			source_ptr++;
+		}
+	}
+}
+
+static void eeprom_xec_wait_transfer_compl(struct eeprom_xec_regs * const regs)
+{
+	uint8_t sts = 0;
+
+	do {
+		sts = XEC_EEPROM_STS_TRANSFER_COMPL & regs->status;
+	} while (!sts);
+}
+
+static void eeprom_xec_wait_write_compl(struct eeprom_xec_regs * const regs)
+{
+	uint8_t sts = 0;
+	uint8_t retry_count = 0;
+
+	do {
+		if (retry_count >= XEC_EEPROM_XFER_COMPL_RETRY_COUNT) {
+			LOG_ERR("XEC EEPROM retry count exceeded");
+			break;
+		}
+
+		regs->buffer[0] = 0;
+
+		/* Issue the READ_STS command */
+		regs->execute = XEC_EEPROM_EXC_CMD_READ_STS;
+
+		k_sleep(K_USEC(XEC_EEPROM_DELAY_US));
+
+		eeprom_xec_wait_transfer_compl(regs);
+
+		/* Clear the appropriate status bits */
+		regs->status = XEC_EEPROM_STS_TRANSFER_COMPL;
+
+		sts = regs->buffer[0] & (XEC_EEPROM_STS_BYTE_WIP |
+							XEC_EEPROM_STS_BYTE_WENB);
+
+		retry_count++;
+
+	} while (sts);
+}
+
+static void eeprom_xec_data_read_32_bytes(struct eeprom_xec_regs * const regs,
+							uint8_t *buf, size_t len, off_t offset)
+{
+	/* Issue the READ command to transfer buffer to EEPROM memory */
+	eeprom_xec_execute_reg_set(regs, len, XEC_EEPROM_EXC_CMD_READ, offset);
+
+	k_sleep(K_USEC(XEC_EEPROM_DELAY_US));
+
+	/* Wait until the read operation has completed */
+	eeprom_xec_wait_transfer_compl(regs);
+
+	/* Clear the appropriate status bits */
+	regs->status = XEC_EEPROM_STS_TRANSFER_COMPL;
```

Comment:
Instead of adding delay `k_sleep(K_USEC(XEC_EEPROM_DELAY_US));` in this (and other) function, if you wrap it inside `eeprom_xec_wait_transfer_compl(regs)` function, you won't have to call the delay separately. 
That way, giving better justice to the function name _xec__**wait**_tranfer as well.

Additionally, it would make sense to also update the `regs->status `within the function `eeprom_xec_wait_transfer_compl`.